src/main/java/org/openstreetmap/atlas/geography/atlas/change/ChangeAtlas.java

```diff
@@ -321,26 +368,23 @@ public Point point(final long identifier)
     @Override
     public Relation relation(final long identifier)
     {
-        final Relation relation = entityFor(identifier, ItemType.RELATION,
-                () -> this.source.relation(identifier),
-                (sourceEntity, overrideEntity) -> new ChangeRelation(this, (Relation) sourceEntity,
-                        (Relation) overrideEntity));
-
         /*
          * If the relation was not found in this atlas, return null. Additionally, we check to see
          * if the relation has no members. If so, it is considered empty and is dropped from the
          * atlas. This logic, combined with the logic in ChangeRelation.membersFor, will
          * automatically handle removing non-empty but shallow relations as well.
          */
-        if (relation == null)
-        {
-            return null;
-        }
-        if (relation.members().isEmpty())
-        {
-            return null;
-        }
-        return relation;
+        final Predicate<ChangeRelation> nullableRelation = relationCandidate -> relationCandidate
+                .members().isEmpty();
+
+        final Supplier<ChangeRelation> creator = () -> entityFor(identifier, ItemType.RELATION,
+                () -> this.source.relation(identifier),
+                (sourceEntity, overrideEntity) -> new ChangeRelation(this, (Relation) sourceEntity,
+                        (Relation) overrideEntity));
+
+        return getFromCacheOrCreate(this.relationsCache, cache -> this.relationsCache = cache,
```

Comment:
I assume that returning a cached `ChangeRelation` on re-requests significantly speeds up the validation step. Is there any significant memory impact? Since validation touches all relations, you are going to have a full cache after validation I would imagine. Unless I am missing something.

Reply:
Yes, full cache after validation, which means each feature is duplicated every time it is accessed. Here they are all duplicated at the end of validation.